tools/gbagfx/jasc_pal.c

```diff
@@ -2,155 +2,87 @@
+    char line_buffer[MAX_LINE_LENGTH];
+    int red, green, blue;
+    int numColors;
 
-    ReadJascPaletteLine(fp, line);
+    FILE *fp = fopen(path, "r");
 
-    if (strcmp(line, "JASC-PAL") != 0)
-        FATAL_ERROR("Invalid JASC-PAL signature.\n");
+    if (!fp)
+        FATAL_ERROR("Cannot open JASC-PAL file \"%s\" with error: %s\n", path, strerror(errno));
 
-    ReadJascPaletteLine(fp, line);
+    // Check JASC-PAL Header
+    ReadJascPaletteLine(fp, line_buffer);
+    if (strcmp(line_buffer, "JASC-PAL") != 0)
+        FATAL_ERROR("Invalid signature, expected \"JASC-PAL\", read: \"%s\"\n",line_buffer);
 
-    if (strcmp(line, "0100") != 0)
+    ReadJascPaletteLine(fp, line_buffer);
+    if (strcmp(line_buffer, "0100") != 0)
         FATAL_ERROR("Unsuported JASC-PAL version.\n");
 
-    ReadJascPaletteLine(fp, line);
+    // Get number of colors in palette
+    ReadJascPaletteLine(fp, line_buffer);
+    numColors = strtol(line_buffer, NULL, 10);
 
-    if (!ParseNumber(line, NULL, 10, &palette->numColors))
-        FATAL_ERROR("Failed to parse number of colors.\n");
-
-    if (palette->numColors < 1 || palette->numColors > 256)
-        FATAL_ERROR("%d is an invalid number of colors. The number of colors must be in the range [1, 256].\n", palette->numColors);
-
-    for (int i = 0; i < palette->numColors; i++)
+    // Check for sensible number of colors
+    if (numColors > 0 && numColors <= 256)
     {
-        ReadJascPaletteLine(fp, line);
-
-        char *s = line;
-        char *end;
-
-        int red;
-        int green;
-        int blue;
-
-        if (!ParseNumber(s, &end, 10, &red))
-            FATAL_ERROR("Failed to parse red color component.\n");
-
-        s = end;
-
-        if (*s != ' ')
-            FATAL_ERROR("Expected a space after red color component.\n");
-
-        s++;
-
-        if (*s < '0' || *s > '9')
-            FATAL_ERROR("Expected only a space between red and green color components.\n");
-
-        if (!ParseNumber(s, &end, 10, &green))
-            FATAL_ERROR("Failed to parse green color component.\n");
-
-        s = end;
-
-        if (*s != ' ')
-            FATAL_ERROR("Expected a space after green color component.\n");
-
-        s++;
-
-        if (*s < '0' || *s > '9')
-            FATAL_ERROR("Expected only a space between green and blue color components.\n");
-
-        if (!ParseNumber(s, &end, 10, &blue))
-            FATAL_ERROR("Failed to parse blue color component.\n");
-
-        if (*end != 0)
-            FATAL_ERROR("Garbage after blue color component.\n");
-
-        if (red < 0 || red > 255)
-            FATAL_ERROR("Red color component (%d) is outside the range [0, 255].\n", red);
-
-        if (green < 0 || green > 255)
-            FATAL_ERROR("Green color component (%d) is outside the range [0, 255].\n", green);
+        palette->numColors = numColors;
+    } else {
+        FATAL_ERROR("Out-of-Bounds number of colors specifed: %i. Maximum supported is 256\n", numColors);
```

Comment:
The original error message was more descriptive:
`"%d is an invalid number of colors. The number of colors must be in the range [1, 256].\n"`

Reply:
Hm ok I'll put the old one in